### PR TITLE
resolved: support reloading configuration at runtime

### PR DESCRIPTION
**Pull request by bluca**
Drop connections and caches and reload config from files, to allow
for low-interruptions updates, and hook up to the usual SIGHUP and
ExecReload=. Mark servers and services configured directly via D-Bus
so that they can be kept around, and only the configuration file
settings are dropped and reloaded.

Fixes https://github.com/systemd/systemd/issues/17503
Fixes https://github.com/systemd/systemd/issues/20604

### CODE REVIEW

File: src/resolve/resolved-dns-server.h [hunk 1]
```diff
@@ -100,6 +100,9 @@ struct DnsServer {
         /* If linked is set, then this server appears in the servers linked list */
         bool linked:1;
         LIST_FIELDS(DnsServer, servers);
+
+        /* Set when the server is configured manually via D-Bus, to avoid removing it on reload */
+        bool from_dbus:1;
```
Comment (yuwata) [inline src/resolve/resolved-dns-server.h]: Please reorder elements with bitfield specifier `:1` to pack them more memory efficient.

Comment (yuwata) [reply src/resolve/resolved-dns-server.h]: Or, maybe, better to make it enum, to explain the source of the server:
```
DNS_SERVER_FROM_CONFIG_FILE,
DNS_SERVER_FROM_NETWORKD,
DNS_SERVER_FROM_DBUS,
```

File: src/resolve/resolved-manager.c [hunk 2]
```diff
@@ -565,6 +566,69 @@ static int manager_memory_pressure_listen(Manager *m) {
...
+
+        (void) sd_notifyf(/* unset= */ false,
+                          "RELOADING=1\n"
+                          "STATUS=Reloading configuration...\n"
+                          "MONOTONIC_USEC=" USEC_FMT, now(CLOCK_MONOTONIC));
+
+        m->llmnr_support = DEFAULT_LLMNR_MODE;
+        m->mdns_support = DEFAULT_MDNS_MODE;
+        m->dnssec_mode = DEFAULT_DNSSEC_MODE;
+        m->dns_over_tls_mode = DEFAULT_DNS_OVER_TLS_MODE;
```
Comment (yuwata) [inline src/resolve/resolved-manager.c:575]: IIRC, these mode can be also changed through DBus. Maybe we should keep them if they are changed by DBus?

**Review by yuwata — CHANGES_REQUESTED** [review]

File: src/resolve/resolved-manager.c (hunk 2 above)
Comment (bluca) [reply src/resolve/resolved-manager.c:575]: Only the per-interface modes can be set via D-Bus, the "global" ones cannot, they are only configured via the config file.

Comment (yuwata) [reply src/resolve/resolved-manager.c:575]: Ah, indeed. Thanks.

File: src/resolve/resolved-manager.c [hunk 3]
```diff
@@ -565,6 +566,68 @@ static int manager_memory_pressure_listen(Manager *m) {
...
+        m->stale_retention_usec = 0;
+
+        dns_server_unlink_on_reload(m->dns_servers);
+        dns_server_unlink_on_reload(m->fallback_dns_servers);
+        m->dns_extra_stub_listeners = ordered_set_free(m->dns_extra_stub_listeners);
+        HASHMAP_FOREACH(dnssd, m->dnssd_services)
+                if (dnssd->config_source == RESOLVE_CONFIG_SOURCE_FILE) {
+                        hashmap_remove(m->dnssd_services, dnssd->name);
+                        dnssd_service_free(dnssd);
+                }
```
Comment (yuwata) [inline src/resolve/resolved-manager.c]: Maybe, better to introduce `dnssd_service_clear_on_reload()` or something?

File: src/resolve/resolved-manager.c [hunk 4]
```diff
@@ -565,6 +566,68 @@ static int manager_memory_pressure_listen(Manager *m) {
...
+        if (r < 0)
+                log_warning_errno(r, "Failed to parse config file, using defaults: %m");
+        else
+                log_info("Config file reloaded.");
+
+        r = dnssd_load(m);
+        if (r < 0)
+                log_warning_errno(r, "Failed to load DNS-SD configuration files: %m");
+
+        r = dns_scope_new(m, &m->unicast_scope, NULL, DNS_PROTOCOL_DNS, AF_UNSPEC);
```
Comment (yuwata) [inline src/resolve/resolved-manager.c:616]: Why is this necessary? Maybe, better to add a short comment?

File: src/resolve/resolved-manager.c [hunk 5]
```diff
@@ -565,6 +566,68 @@ static int manager_memory_pressure_listen(Manager *m) {
...
+        if (r < 0)
+                return r;
+
+        r = on_network_event(/* sd_event_source= */ NULL, -EBADF, /* revents= */ 0, m);
```
Comment (yuwata) [inline src/resolve/resolved-manager.c:622]: Also here, why? IIUC, in the above we clear settings given by resolved.conf, and it provides global setting. But, this provides per-link settings. I do not think this is necessary. If it is really necessary, please add a short comment.

**Review by yuwata — CHANGES_REQUESTED** [review]
Looks mostly good. But, several questions.

File: src/resolve/resolved-manager.c (hunk 4 above)
Comment (bluca) [reply src/resolve/resolved-manager.c:616]: It's influenced by the config file, so it's reloaded to take any changes into account, same as below. Added comments.

File: src/resolve/resolved-dns-server.h [hunk 6]
```diff
@@ -100,6 +102,9 @@ struct DnsServer {
         /* If linked is set, then this server appears in the servers linked list */
         bool linked:1;
         LIST_FIELDS(DnsServer, servers);
+
+        /* Server registered via D-Bus are not removed on reload */
```
Comment (keszybz) [inline src/resolve/resolved-dns-server.h]: Server**s**

**Review by yuwata — APPROVED** [review]
LGTM.

File: src/resolve/resolved-manager.c [hunk 7]
```diff
@@ -565,6 +566,67 @@ static int manager_memory_pressure_listen(Manager *m) {
...
+        m->llmnr_support = DEFAULT_LLMNR_MODE;
+        m->mdns_support = DEFAULT_MDNS_MODE;
+        m->dnssec_mode = DEFAULT_DNSSEC_MODE;
+        m->dns_over_tls_mode = DEFAULT_DNS_OVER_TLS_MODE;
+        m->enable_cache = DNS_CACHE_MODE_YES;
+        m->dns_stub_listener_mode = DNS_STUB_LISTENER_YES;
+        m->read_etc_hosts = true;
+        m->resolve_unicast_single_label = false;
+        m->cache_from_localhost = false;
+        m->stale_retention_usec = 0;
```
Comment (keszybz) [inline src/resolve/resolved-manager.c:581]: This part overlaps in functionality with `manager_new()`. I really don't like having this code twice in two different forms. Can we at least extract `manager_set_defaults(Manager *m)` that will do this part?

File: src/resolve/resolved-manager.c [hunk 8]
```diff
@@ -565,6 +566,67 @@ static int manager_memory_pressure_listen(Manager *m) {
...
+
+        dns_trust_anchor_flush(&m->trust_anchor);
+
+        r = dns_trust_anchor_load(&m->trust_anchor);
+        if (r < 0)
+                return r;
+
+        r = manager_parse_config_file(m);
+        if (r < 0)
+                log_warning_errno(r, "Failed to parse config file, using defaults: %m");
```
Comment (keszybz) [inline src/resolve/resolved-manager.c]: Hmm, I don't think this message is correct. `manager_parse_config_file()` is not atomic, so we may well be using a mix of defaults and updated settings.

File: src/resolve/resolved-manager.c [hunk 9]
```diff
@@ -565,6 +566,67 @@ static int manager_memory_pressure_listen(Manager *m) {
...
+        else
+                log_info("Config file reloaded.");
+
+        r = dnssd_load(m);
+        if (r < 0)
+                log_warning_errno(r, "Failed to load DNS-SD configuration files: %m");
+
+        /* The default scope configuration is influenced by the manager's configuration (modes, etc.) so
```
Comment (keszybz) [inline src/resolve/resolved-manager.c]: Comma missing before "so".

**Review by keszybz — CHANGES_REQUESTED** [review]

**Review by keszybz — APPROVED** [review]

**Comment by fxzxmicah** [comment]
Sorry to bother you, may I ask if `reload` supports reloading configuration files at `/etc/systemd/resolved.conf.d/*.conf`?
